File: czsc/signals/coo.py
# -*- coding: utf-8 -*-
"""
author: zengbin93
email: zeng_bin8888@163.com
create_dt: 2022/11/10 23:14
describe: coo 是 cooperation 的缩写，作为前缀代表信号开源协作成员贡献的信号
"""
import numpy as np
from deprecated import deprecated
from collections import OrderedDict
from czsc import CZSC
from czsc.utils import create_single_signal, get_sub_elements
from czsc.signals.tas import update_ma_cache, update_sar_cache, update_kdj_cache, update_cci_cache


def __cal_td_seq(close: np.ndarray):
    """TDSEQ计算辅助函数

    正值表示上涨，负值表示下跌

    :param close: np.array
        收盘价序列
    :return: np.array
    """
    if len(close) < 5:
        return np.zeros(len(close), dtype=np.int32)

    res = np.zeros(len(close), dtype=np.int32)
    for i in range(4, len(close)):
        if close[i] > close[i - 4]:
            res[i] = res[i - 1] + 1
        elif close[i] < close[i - 4]:
            res[i] = res[i - 1] - 1

    return res


@deprecated(version='1.0.0', reason="请使用 coo_td_V221111")
def coo_td_V221110(c: CZSC, **kwargs) -> OrderedDict:
    """获取倒数第i根K线的TD信号

    参数模板："{freq}_D{di}K_TD"

    **信号列表：**

    - Signal('60分钟_D2K_TD_延续_非底_任意_0')
    - Signal('60分钟_D2K_TD_延续_非顶_任意_0')
    - Signal('60分钟_D2K_TD_延续_TD顶_任意_0')
    - Signal('60分钟_D2K_TD_看空_非底_任意_0')
    - Signal('60分钟_D2K_TD_延续_TD底_任意_0')
    - Signal('60分钟_D2K_TD_看多_非顶_任意_0')

    :param c: CZSC对象
    :param di: 倒数第di根K线
    :return: s
    """
    di = int(kwargs.get("di", 1))
    k1, k2, k3 = f"{c.freq.value}_D{di}K_TD".split("_")

    if di == 1:
        close = np.array([x.close for x in c.bars_raw[-50:]])
    else:
        close = np.array([x.close for x in c.bars_raw[-50 - di + 1 : -di + 1]])

    td = __cal_td_seq(close)
    if td[-1] > 0:
        v1 = '看多' if len(td) > 1 and td[-2] < -8 else '延续'
        v2 = 'TD顶' if td[-1] > 8 else '非顶'
    elif td[-1] < 0:
        v1 = '看空' if len(td) > 1 and td[-2] > 8 else '延续'
        v2 = 'TD底' if td[-1] < -8 else '非底'
    else:
        v1 = '其他'
        v2 = '其他'

    return create_single_signal(k1=k1, k2=k2, k3=k3, v1=v1, v2=v2)


def coo_td_V221111(c: CZSC, **kwargs) -> OrderedDict:
    """获取倒数第i根K线的TD信号

    参数模板："{freq}_D{di}TD_BS辅助V221111"

    **信号逻辑：**

    神奇九转指标

    **信号列表：**

    - Signal('15分钟_D1TD_BS辅助V221111_延续_TD底_任意_0')
    - Signal('15分钟_D1TD_BS辅助V221111_看多_非顶_任意_0')
    - Signal('15分钟_D1TD_BS辅助V221111_延续_非顶_任意_0')
    - Signal('15分钟_D1TD_BS辅助V221111_延续_非底_任意_0')
    - Signal('15分钟_D1TD_BS辅助V221111_延续_TD顶_任意_0')
    - Signal('15分钟_D1TD_BS辅助V221111_看空_非底_任意_0')

    :param c: CZSC对象
    :param di: 倒数第di根K线
    :return: s
    """
    di = int(kwargs.get("di", 1))
    freq = c.freq.value
    k1, k2, k3 = f"{freq}_D{di}TD_BS辅助V221111".split("_")
    if len(c.bars_raw) < 50 + di:
        return create_single_signal(k1=k1, k2=k2, k3=k3, v1="其他")

    bars = get_sub_elements(c.bars_raw, di=di, n=50)
    close = np.array([x.close for x in bars])
    td = __cal_td_seq(close)

    if td[-1] > 0:
        v1 = '看多' if len(td) > 1 and td[-2] < -8 else '延续'
        v2 = 'TD顶' if td[-1] > 8 else '非顶'
    elif td[-1] < 0:
        v1 = '看空' if len(td) > 1 and td[-2] > 8 else '延续'
        v2 = 'TD底' if td[-1] < -8 else '非底'
    else:
        v1 = '其他'
        v2 = '其他'

    return create_single_signal(k1=k1, k2=k2, k3=k3, v1=v1, v2=v2)


def coo_cci_V230323(c: CZSC, **kwargs) -> OrderedDict:
    """CCI结合均线的多空信号

    参数模板："{freq}_D{di}CCI{n}#{ma_type}#{m}_BS辅助V230323"

     **信号逻辑：**

    1. CCI大于100，且向上突破均线，看多；
    2. CCI小于-100，且向下突破均线，看空；

    **信号列表：**

    - Signal('15分钟_D1CCI20#SMA#5_BS辅助V230323_空头_向下_任意_0')
    - Signal('15分钟_D1CCI20#SMA#5_BS辅助V230323_空头_向上_任意_0')
    - Signal('15分钟_D1CCI20#SMA#5_BS辅助V230323_多头_向上_任意_0')
    - Signal('15分钟_D1CCI20#SMA#5_BS辅助V230323_多头_向下_任意_0')

    :param c: CZSC对象
    :param kwargs: 参数字典
        - :param di: 信号计算截止倒数第i根K线
        - :param n: CCI的计算周期
        - :param m: 乘以N表示均线的计算周期
    :return: 返回信号结果
    """
    di = int(kwargs.get("di", 1))
    n = int(kwargs.get("n", 20))
    m = int(kwargs.get("m", 5))
    ma_type = kwargs.get('ma_type', 'SMA').upper()
    freq = c.freq.value
    cache_key_cci = update_cci_cache(c, timeperiod=n)
    cache_key_ma = update_ma_cache(c, ma_type=ma_type, timeperiod=n * m)

    k1, k2, k3 = f"{freq}_D{di}CCI{n}#{ma_type}#{m}_BS辅助V230323".split('_')
    v1 = '其他'
    if len(c.bars_raw) < n * m + di:
        return create_single_signal(k1=k1, k2=k2, k3=k3, v1=v1)

    bars = get_sub_elements(c.bars_raw, di=di, n=2)
    cci = bars[-1].cache[cache_key_cci]
    MA_CC = bars[-1].cache[cache_key_ma]

    if cci > 100 and bars[-1].close > MA_CC:
        v1 = '多头'

    if cci < -100 and bars[-1].close < MA_CC:
        v1 = '空头'

    if v1 == '其他':
        return create_single_signal(k1=k1, k2=k2, k3=k3, v1=v1)

    v2 = "向上" if cci >= bars[-2].cache[cache_key_cci] else "向下"
    return create_single_signal(k1=k1, k2=k2, k3=k3, v1=v1, v2=v2)


def coo_kdj_V230322(c: CZSC, **kwargs) -> OrderedDict:
    """均线判定方向，KD决定进场时机

    参数模板："{freq}_D{di}KDJ{fastk_period}#{slowk_period}#{slowd_period}#{ma_type}#{n}_BS辅助V230322"

     **信号逻辑：**

     1. K线向上突破均线，且 K < D，看多；
     2. K线向下突破均线，且 K > D，看空；

     **信号列表：**

    - Signal('15分钟_D1KDJ9#3#3#EMA#20_BS辅助V230322_空头_任意_任意_0')
    - Signal('15分钟_D1KDJ9#3#3#EMA#20_BS辅助V230322_多头_任意_任意_0')

    :param c: CZSC对象
    :param kwargs: 参数字典
        - :param di: 信号计算截止倒数第i根K线
        - :param n: 均线计算周期
        - :param fastk_period:  kdj参数
        - :param slowk_period:  kdj参数
        - :param slowd_period:  kdj参数
     :return: 返回信号结果
    """
    di = int(kwargs.get("di", 1))
    n = int(kwargs.get("n", 3))
    ma_type = kwargs.get('ma_type', 'EMA').upper()
    fastk_period = int(kwargs.get('fastk_period', 9))
    slowk_period = int(kwargs.get('slowk_period', 3))
    slowd_period = int(kwargs.get('slowd_period', 3))

    ma = update_ma_cache(c, ma_type=ma_type, timeperiod=n)
    cache_key = update_kdj_cache(c, fastk_period=fastk_period, slowk_period=slowk_period, slowd_period=slowd_period)

    freq = c.freq.value
    k1, k2, k3 = f"{freq}_D{di}KDJ{fastk_period}#{slowk_period}#{slowd_period}#{ma_type}#{n}_BS辅助V230322".split('_')
    if len(c.bars_raw) < fastk_period * slowk_period + di:
        return create_single_signal(k1=k1, k2=k2, k3=k3, v1="其他")

    _bars = get_sub_elements(c.bars_raw, di=di, n=n)
    kdj, mac = _bars[-1].cache[cache_key], _bars[-1].cache[ma]

    if _bars[-1].close > mac and kdj['k'] < kdj['d']:
        v1 = "多头"
    elif _bars[-1].close < mac and kdj['k'] > kdj['d']:
        v1 = "空头"
    else:
        v1 = "其他"

    return create_single_signal(k1=k1, k2=k2, k3=k3, v1=v1)


def coo_sar_V230325(c: CZSC, **kwargs) -> OrderedDict:
    """SAR和高低点结合判断买卖时机

    参数模板："{freq}_D{di}N{n}SAR_BS辅助V230325"

     **信号逻辑：**

     1. 最高价大于N个周期收盘价的最高价，收盘价在SAR上方，看多；
     2. 最低价小于N个周期收盘价的最低价，收盘价在SAR下方，看空；

     **信号列表：**

    - Signal('15分钟_D1N20SAR_BS辅助V230325_空头_任意_任意_0')
    - Signal('15分钟_D1N20SAR_BS辅助V230325_多头_任意_任意_0')

    :param c: CZSC对象
    :param kwargs: 参数字典
        - :param di: 信号计算截止倒数第i根K线
        - :param n: 信号计算的K线数量
     :return: 返回信号结果
    """
    di = int(kwargs.get("di", 1))
    n = int(kwargs.get("n", 60))
    cache_key = update_sar_cache(c)

    freq = c.freq.value
    k1, k2, k3 = f"{freq}_D{di}N{n}SAR_BS辅助V230325".split('_')
    v1 = "其他"
    if len(c.bars_raw) < n + di + 10:
        return create_single_signal(k1=k1, k2=k2, k3=k3, v1=v1)

    _bars = get_sub_elements(c.bars_raw, di=di, n=n)
    hhv = max([x.close for x in _bars])
    llv = min([x.close for x in _bars])
    sar, close = _bars[-1].cache[cache_key], _bars[-1].close

    if close > sar and _bars[-1].high >= hhv:
        v1 = "多头"
    if close < sar and _bars[-1].low <= llv:
        v1 = "空头"

    return create_single_signal(k1=k1, k2=k2, k3=k3, v1=v1)

File: crates/czsc-signals/src/coo.rs
use crate::params::ParamView;
use crate::types::TaCache;
use crate::utils::sig::{get_sub_elements, make_kline_signal_v1, make_kline_signal_v2};
use crate::utils::ta::{update_cci_cache, update_kdj_cache, update_ma_cache, update_sar_cache};
use czsc_core::analyze::CZSC;
use czsc_core::objects::signal::Signal;
use czsc_signal_macros::signal;

fn cal_td_seq(close: &[f64]) -> Vec<i32> {
    if close.len() < 5 {
        return vec![0; close.len()];
    }
    let mut res = vec![0; close.len()];
    for i in 4..close.len() {
        if close[i] > close[i - 4] {
            res[i] = res[i - 1] + 1;
        } else if close[i] < close[i - 4] {
            res[i] = res[i - 1] - 1;
        }
    }
    res
}

fn td_signal_from_close(close: &[f64]) -> (&'static str, &'static str) {
    let td = cal_td_seq(close);
    let x = *td.last().unwrap_or(&0);
    if x > 0 {
        let v1 = if td.len() > 1 && td[td.len() - 2] < -8 {
            "看多"
        } else {
            "延续"
        };
        let v2 = if x > 8 { "TD顶" } else { "非顶" };
        (v1, v2)
    } else if x < 0 {
        let v1 = if td.len() > 1 && td[td.len() - 2] > 8 {
            "看空"
        } else {
            "延续"
        };
        let v2 = if x < -8 { "TD底" } else { "非底" };
        (v1, v2)
    } else {
        ("其他", "其他")
    }
}

/// coo_td_V221110：TD 神奇九转信号（旧版模板）
///
/// 参数模板：`"{freq}_D{di}K_TD"`
///
/// 信号逻辑：
/// 1. 取倒数 `di` 截止的最近 50 根收盘价；
/// 2. 按 `close[i]` 与 `close[i-4]` 比较累计 TD 计数；
/// 3. 根据最新 TD 值及前一值输出 `看多/看空/延续` 与 `TD顶/TD底/非顶/非底`。
///
/// 信号列表示例：
/// - `Signal('60分钟_D1K_TD_延续_非顶_任意_0')`
/// - `Signal('60分钟_D1K_TD_看空_TD底_任意_0')`
///
/// 参数说明：
/// - `di`：信号计算截止在倒数第 `di` 根K线，默认 `1`。
/// 对齐说明：与 Python `coo_td_V221110` 的 TD 计数递推一致。
#[signal(
    category = "kline",
    name = "coo_td_V221110",
    template = "{freq}_D{di}K_TDV221110",
    opcode = "CooTdV221110",
    param_kind = "CooTdV221110"
)]
pub fn coo_td_v221110(c: &CZSC, params: &ParamView, _cache: &mut TaCache) -> Vec<Signal> {
    let di = params.usize("di", 1);
    let k1 = c.freq.to_string();
    let k2 = format!("D{}K", di);
    let k3 = "TD";
    let bars = get_sub_elements(&c.bars_raw, di, 50);
    if bars.is_empty() {
        return make_kline_signal_v1(&k1, &k2, k3, "其他");
    }
    let close: Vec<f64> = bars.iter().map(|x| x.close).collect();
    let (v1, v2) = td_signal_from_close(&close);
    make_kline_signal_v2(&k1, &k2, k3, v1, v2)
}

/// coo_td_V221111：TD 神奇九转信号
///
/// 参数模板：`"{freq}_D{di}TD_BS辅助V221111"`
///
/// 信号逻辑：
/// 1. 取倒数 `di` 截止的最近 50 根收盘价；
/// 2. 计算 TD 计数序列；
/// 3. 输出 `看多/看空/延续` 与 `TD顶/TD底/非顶/非底` 组合。
///
/// 信号列表示例：
/// - `Signal('60分钟_D1TD_BS辅助V221111_延续_非顶_任意_0')`
/// - `Signal('60分钟_D1TD_BS辅助V221111_看多_TD顶_任意_0')`
///
/// 参数说明：
/// - `di`：信号计算截止在倒数第 `di` 根K线，默认 `1`。
/// 对齐说明：与 Python `coo_td_V221111` 的窗口和判定分支一致。
#[signal(
    category = "kline",
    name = "coo_td_V221111",
    template = "{freq}_D{di}TD_BS辅助V221111",
    opcode = "CooTdV221111",
    param_kind = "CooTdV221111"
)]
pub fn coo_td_v221111(c: &CZSC, params: &ParamView, _cache: &mut TaCache) -> Vec<Signal> {
    let di = params.usize("di", 1);
    let k1 = c.freq.to_string();
    let k2 = format!("D{}TD", di);
    let k3 = "BS辅助V221111";
    if c.bars_raw.len() < 50 + di {
        return make_kline_signal_v1(&k1, &k2, k3, "其他");
    }
    let bars = get_sub_elements(&c.bars_raw, di, 50);
    let close: Vec<f64> = bars.iter().map(|x| x.close).collect();
    let (v1, v2) = td_signal_from_close(&close);
    make_kline_signal_v2(&k1, &k2, k3, v1, v2)
}

/// coo_cci_V230323：CCI 结合均线的多空与方向信号
///
/// 参数模板：`"{freq}_D{di}CCI{n}#{ma_type}#{m}_BS辅助V230323"`
///
/// 信号逻辑：
/// 1. 计算 `CCI(n)` 与 `MA(n*m)`；
/// 2. `CCI>100` 且 `close>MA` 判 `多头`，`CCI<-100` 且 `close<MA` 判 `空头`；
/// 3. 若非 `其他`，再比较前后两根 CCI 输出 `向上/向下`。
///
/// 信号列表示例：
/// - `Signal('60分钟_D1CCI20#SMA#5_BS辅助V230323_多头_向上_任意_0')`
/// - `Signal('60分钟_D1CCI20#SMA#5_BS辅助V230323_空头_向下_任意_0')`
///
/// 参数说明：
/// - `di`：信号计算截止在倒数第 `di` 根K线，默认 `1`；
/// - `n`：CCI 周期，默认 `20`；
/// - `m`：均线倍数（最终周期 `n*m`），默认 `5`；
/// - `ma_type`：均线类型，默认 `SMA`。
/// 对齐说明：与 Python `coo_cci_V230323` 的阈值和方向判定一致。
#[signal(
    category = "kline",
    name = "coo_cci_V230323",
    template = "{freq}_D{di}CCI{n}#{ma_type}#{m}_BS辅助V230323",
    opcode = "CooCciV230323",
    param_kind = "CooCciV230323"
)]
pub fn coo_cci_v230323(c: &CZSC, params: &ParamView, cache: &mut TaCache) -> Vec<Signal> {
    let di = params.usize("di", 1);
    let n = params.usize("n", 20);
    let m = params.usize("m", 5);
    let ma_type = params.str("ma_type", "SMA").to_uppercase();

    let cci_key = format!("CCI{}", n);
    let ma_key = format!("{}#{}", ma_type, n * m);
    update_cci_cache(c, &cci_key, n, cache);
    update_ma_cache(c, &ma_key, &ma_type, n * m, cache);

    let k1 = c.freq.to_string();
    let k2 = format!("D{}CCI{}#{}#{}", di, n, ma_type, m);
    let k3 = "BS辅助V230323";
    if c.bars_raw.len() < n * m + di {
        return make_kline_signal_v1(&k1, &k2, k3, "其他");
    }
    if c.bars_raw.len() < di + 1 {
        return make_kline_signal_v1(&k1, &k2, k3, "其他");
    }

    let idx = c.bars_raw.len() - di;
    let prev_idx = idx.saturating_sub(1);
    let cci = cache
        .series
        .get(&cci_key)
        .and_then(|x| x.get(idx))
        .copied()
        .unwrap_or(f64::NAN);
    let cci_prev = cache
        .series
        .get(&cci_key)
        .and_then(|x| x.get(prev_idx))
        .copied()
        .unwrap_or(f64::NAN);
    let ma = cache
        .series
        .get(&ma_key)
        .and_then(|x| x.get(idx))
        .copied()
        .unwrap_or(f64::NAN);
    let close = c.bars_raw[idx].close;

    let mut v1 = "其他";
    if cci > 100.0 && close > ma {
        v1 = "多头";
    }
    if cci < -100.0 && close < ma {
        v1 = "空头";
    }
    if v1 == "其他" {
        return make_kline_signal_v1(&k1, &k2, k3, v1);
    }
    let v2 = if cci >= cci_prev { "向上" } else { "向下" };
    make_kline_signal_v2(&k1, &k2, k3, v1, v2)
}

/// coo_kdj_V230322：均线与 KDJ 配合多空信号
///
/// 参数模板：`"{freq}_D{di}KDJ{fastk_period}#{slowk_period}#{slowd_period}#{ma_type}#{n}_BS辅助V230322"`
///
/// 信号逻辑：
/// 1. 计算 `KDJ` 与 `MA(n)`；
/// 2. `close > MA` 且 `K < D` 判 `多头`；
/// 3. `close < MA` 且 `K > D` 判 `空头`，否则 `其他`。
///
/// 信号列表示例：
/// - `Signal('60分钟_D1KDJ9#3#3#EMA#3_BS辅助V230322_多头_任意_任意_0')`
/// - `Signal('60分钟_D1KDJ9#3#3#EMA#3_BS辅助V230322_空头_任意_任意_0')`
///
/// 参数说明：
/// - `di`：信号计算截止在倒数第 `di` 根K线，默认 `1`；
/// - `n`：均线周期，默认 `3`；
/// - `ma_type`：均线类型，默认 `EMA`；
/// - `fastk_period/slowk_period/slowd_period`：KDJ 参数，默认 `9/3/3`。
/// 对齐说明：与 Python `coo_kdj_V230322` 的组合条件一致。
#[signal(
    category = "kline",
    name = "coo_kdj_V230322",
    template = "{freq}_D{di}KDJ{fastk_period}#{slowk_period}#{slowd_period}#{ma_type}#{n}_BS辅助V230322",
    opcode = "CooKdjV230322",
    param_kind = "CooKdjV230322"
)]
pub fn coo_kdj_v230322(c: &CZSC, params: &ParamView, cache: &mut TaCache) -> Vec<Signal> {
    let di = params.usize("di", 1);
    let n = params.usize("n", 3);
    let ma_type = params.str("ma_type", "EMA").to_uppercase();
    let fastk_period = params.usize("fastk_period", 9);
    let slowk_period = params.usize("slowk_period", 3);
    let slowd_period = params.usize("slowd_period", 3);

    let ma_key = format!("{}#{}", ma_type, n);
    let kdj_key = format!("KDJ{}#{}#{}", fastk_period, slowk_period, slowd_period);
    update_ma_cache(c, &ma_key, &ma_type, n, cache);
    update_kdj_cache(
        c,
        &kdj_key,
        fastk_period,
        slowk_period,
        slowd_period,
        cache,
    );

    let k1 = c.freq.to_string();
    let k2 = format!(
        "D{}KDJ{}#{}#{}#{}#{}",
        di, fastk_period, slowk_period, slowd_period, ma_type, n
    );
    let k3 = "BS辅助V230322";
    if c.bars_raw.len() < fastk_period * slowk_period + di {
        return make_kline_signal_v1(&k1, &k2, k3, "其他");
    }
    let idx = c.bars_raw.len() - di;
    let close = c.bars_raw[idx].close;
    let ma = cache
        .series
        .get(&ma_key)
        .and_then(|x| x.get(idx))
        .copied()
        .unwrap_or(f64::NAN);
    let (k, d) = match cache.kdj.get(&kdj_key) {
        Some(kdj) => (
            kdj.k.get(idx).copied().unwrap_or(f64::NAN),
            kdj.d.get(idx).copied().unwrap_or(f64::NAN),
        ),
        None => (f64::NAN, f64::NAN),
    };
    let v1 = if close > ma && k < d {
        "多头"
    } else if close < ma && k > d {
        "空头"
    } else {
        "其他"
    };
    make_kline_signal_v1(&k1, &k2, k3, v1)
}

/// coo_sar_V230325：SAR 与区间极值配合信号
///
/// 参数模板：`"{freq}_D{di}N{n}SAR_BS辅助V230325"`
///
/// 信号逻辑：
/// 1. 计算最近 `n` 根收盘价区间高低点；
/// 2. 若 `close > SAR` 且 `high >= 区间最高收盘` 判 `多头`；
/// 3. 若 `close < SAR` 且 `low <= 区间最低收盘` 判 `空头`。
///
/// 信号列表示例：
/// - `Signal('60分钟_D1N60SAR_BS辅助V230325_多头_任意_任意_0')`
/// - `Signal('60分钟_D1N60SAR_BS辅助V230325_空头_任意_任意_0')`
///
/// 参数说明：
/// - `di`：信号计算截止在倒数第 `di` 根K线，默认 `1`；
/// - `n`：区间窗口，默认 `60`。
/// 对齐说明：与 Python `coo_sar_V230325` 的 SAR 与区间条件一致。
#[signal(
    category = "kline",
    name = "coo_sar_V230325",
    template = "{freq}_D{di}N{n}SAR_BS辅助V230325",
    opcode = "CooSarV230325",
    param_kind = "CooSarV230325"
)]
pub fn coo_sar_v230325(c: &CZSC, params: &ParamView, cache: &mut TaCache) -> Vec<Signal> {
    let di = params.usize("di", 1);
    let n = params.usize("n", 60);
    let sar_key = "SAR".to_string();
    update_sar_cache(c, &sar_key, cache);
    let k1 = c.freq.to_string();
    let k2 = format!("D{}N{}SAR", di, n);
    let k3 = "BS辅助V230325";
    if c.bars_raw.len() < n + di + 10 {
        return make_kline_signal_v1(&k1, &k2, k3, "其他");
    }

    let bars = get_sub_elements(&c.bars_raw, di, n);
    if bars.is_empty() {
        return make_kline_signal_v1(&k1, &k2, k3, "其他");
    }
    let hhv = bars.iter().map(|x| x.close).fold(f64::NEG_INFINITY, f64::max);
    let llv = bars.iter().map(|x| x.close).fold(f64::INFINITY, f64::min);
    let idx = c.bars_raw.len() - di;
    let sar = cache
        .series
        .get(&sar_key)
        .and_then(|x| x.get(idx))
        .copied()
        .unwrap_or(f64::NAN);
    let last = &c.bars_raw[idx];
    let close = last.close;

    let mut v1 = "其他";
    if close > sar && last.high >= hhv {
        v1 = "多头";
    }
    if close < sar && last.low <= llv {
        v1 = "空头";
    }
    make_kline_signal_v1(&k1, &k2, k3, v1)
}
